File: lab3_forecast/gen_aggregate_pos_data.rb
require 'date'
require 'time'

class PointOfSale
  PRODUCTS = [ "staplers", "post-its", "markers", "dry-eraser", "flowers", "beer" ]
  STORES   = [ "SFO1", "SFO2", "SFO3", "LAX", "NYC", "SEA", "BOS" ]

  # simulates a sale at the register
  def sale_event(time:)
    is_weekend = weekend?(time)

    demand  = is_weekend ? rand(100..300) : rand(1..50) # if weekend, simulate high demand
    product = is_weekend ? PRODUCTS[rand(4..5)] : PRODUCTS[rand(0..3)] # choose product depending on weekday/weekend

    {
      timestamp: time.strftime("%Y-%m-%d %H:00:00"), # Amazon Forecast needs yyyy-mm-dd HH:mm:ss format
      item_id:   product,
      demand:    demand,
      location:  STORES[rand(0..STORES.count-1)]
    }
  end

  # generates simulated aggregated sales on an hourly basis
  def aggregate_hourly_sale_events(time:)
    products_sold = rand(1..PRODUCTS.count-1)

    sale_events = []
    products_sold.times do |product|
      sale_events << sale_event(time: time)
    end

    sale_events
  end

  private

    def weekend?(time)
      time.saturday? || time.sunday?
    end
end


# Main

begin
  point_of_sale = PointOfSale.new
  sale_events   = []

  # Generate sales data for 3 months. We will assume stores in different locations
  # operate from 9a - 6p. We will generate aggregate sales by the hour.
  from  = Date.parse("Jul 1, 2019")    # start date
  to    = Date.parse("Sep 30, 2019")   # and end date

  (from..to).each do |date| # for each day between start date and end date...
    (10..17).each do |hour|    # for each hour in that day...
      time = Time.parse("#{date.to_s} #{hour}:00:00")

      hourly_sale_events = point_of_sale.aggregate_hourly_sale_events(time: time)
      hourly_sale_events.each { |sale_event| sale_events << sale_event.values.join(',') }
    end
  end

  # write the list (array) of sale events to a CSV file
  open('retail_analytics.csv', 'w') { |f| f.puts sale_events }

  puts "Done writing events to retail_analytics.csv"
end

File: lab3_forecast/src/gen_aggregate_pos_data.py
import datetime as dt
import random

class PointOfSale:
    PRODUCTS = [ "staplers", "post-its", "markers", "dry-eraser", "flowers", "beer" ]
    STORES   = [ "SFO1", "SFO2", "SFO3", "LAX", "NYC", "SEA", "BOS" ]

    # simulates a sale at the register
    def sale_event(self, time):
        is_weekend = self._is_weekend(time)

        demand  = random.randint(100, 300) if is_weekend else random.randint(1, 50)
        product = self.PRODUCTS[random.randint(4, 5)] if is_weekend else self.PRODUCTS[random.randint(0, 3)]

        return {
            "timestamp": time.strftime('%Y-%m-%d %H:00:00'),
            "item_id": product,
            "demand": demand,
            "location": self.STORES[random.randint(0, len(self.STORES) - 1)]
        }

    # generates simulated aggregated sales on an hourly basis
    def aggregate_hourly_sale_events(self, time):
        no_of_products_sold = random.randint(1, len(self.PRODUCTS) - 1)

        sale_events = []
        for i in range(no_of_products_sold):
            sale_events.append(self.sale_event(time))

        return sale_events

    # private

    def _is_weekend(self, time):
        return time.weekday() > 5


# Main

pos = PointOfSale()
sale_events = []

# Generate sales data for 3 months. We will assume stores in different locations
# operate from 9a - 6p. We will generate aggregate sales by the hour.
start_date = dt.datetime.strptime("Jul 1, 2019", "%b %d, %Y")
end_date   = dt.datetime.strptime("Sep 30, 2019", "%b %d, %Y")

for day in range((end_date - start_date).days + 1): # for each day from start date to end date...
    date = start_date + dt.timedelta(days = day)

    for hour in range(10, 17): # for each working hour in the day...
        time = date + dt.timedelta(hours = hour)
        sale_events += pos.aggregate_hourly_sale_events(time)

f = open('retail_analytics_python.csv', 'w+')

for event in sale_events:
    # the below just converts an event in json dict format to csv
    event_as_csv = ",".join(list(str(value) for value in event.values()))
    print(event_as_csv)
    f.write(event_as_csv + "\n")

print("Done writing events to retail_analytics_python.csv")
